### PR TITLE
[misc] Optimize speculative decoding

### PR DESCRIPTION
**Pull request by jacob-crux**
This PR introduces optimizations to the speculative decoding process by reducing memory copy overhead, resulting in improved performance.

After profiling, I found that cuda memcpy occurred abnormally, so I improved the code. (Refer to the profiling image below)
I experimented with one a100 gpu, and the code I used for the experiment is as follows.

I improved the `_filter_model_output` function and `_merge_outputs` function, and confirmed that the test code below improved by about 3%.

| Applied Function | E2E Latency |
| --- | --- |
| baseline | 0.462s |
| `_filter_model_output` | 0.457s |
| `_merge_outputs` | 0.459s |
| `_filter_model_output` + `_merge_outputs` | 0.449s |

<details>
<summary>Test code using the OPT model</summary>

```python
import time
import nvtx
from typing import List
from vllm import LLM, SamplingParams

def time_generation(llm: LLM, prompts: List[str],
                    sampling_params: SamplingParams):
    llm.generate(prompts, sampling_params)
    llm.generate(prompts, sampling_params)
    times = []
    with nvtx.annotate("generate", color="blue"):
        for _ in range(10):
            start = time.time()
            outputs = llm.generate(prompts, sampling_params)
            end = time.time()
            times.append(end-start)
    print(sum(times) / len(times))

    for output in outputs:
        generated_text = output.outputs[0].text
        print(f"text: {generated_text!r}")


if __name__ == "__main__":
    sampling_params = SamplingParams(temperature=0.0, max_tokens=100)

    prompts = [
        "The president of the United States is",
    ]

    llm = LLM(
        model="facebook/opt-6.7b",
        tensor_parallel_size=1,
        speculative_model="facebook/opt-125m",
        num_speculative_tokens=5,
        use_v2_block_manager=True,
    )

    time_generation(llm, prompts, sampling_params)
```
</details>

**Profiling results**
- Before code improvement
        

- After code improvement

### CODE REVIEW

File: vllm/spec_decode/top1_proposer.py [hunk 1]
```diff
@@ -226,47 +226,31 @@ def _merge_outputs(
...
-            proposal_lens_tensor = torch.tensor(0,
-                                                dtype=torch.long,
-                                                device=self._device).expand(
-                                                    len(proposal_lens))
+            proposal_tokens = torch.full((batch_size, proposal_len), -1,
+                                         dtype=torch.long, device=self._device)
+            proposal_probs = torch.zeros((batch_size, proposal_len, self._vocab_size),
+                                         dtype=torch.float32, device=self._device)
+            proposal_lens_tensor = torch.zeros(len(proposal_lens),
+                                               dtype=torch.long, device=self._device)
```
Comment (njhill) [inline vllm/spec_decode/top1_proposer.py:234]: What's the reason for this change? the existing code minimizes the allocations, I'd be surprised if this was faster.

File: vllm/spec_decode/multi_step_worker.py [hunk 2]
```diff
@@ -185,26 +186,33 @@ def _filter_model_output(
...
+            sampled_token_ids = _filter_optional_tensor(expanded_batch_output.sampled_token_ids, output_indices_to_retain_tensor)
+
+            filtered_outputs.append(SamplerOutput(
+                outputs=outputs,
+                sampled_token_probs=sampled_token_probs,
+                logprobs=logprobs,
+                sampled_token_ids=sampled_token_ids
+            ))
+
+        return filtered_outputs
```
Comment (njhill) [inline vllm/spec_decode/multi_step_worker.py:215]: @jacob-crux I'm not sure the existing code needs to be rewritten, wouldn't it be sufficient to just add this at the start?
```python
        if not expanded_batch_outputs:
            return []

        output_indices_to_retain = torch.tensor(output_indices_to_retain, device=device)
```

File: vllm/spec_decode/top1_proposer.py [hunk 3]
```diff
@@ -226,47 +226,31 @@ def _merge_outputs(
...
-
-        proposal_tokens, proposal_probs = (
-            entire_proposal_tokens,
-            entire_proposal_probs,
-        )
+        proposal_tokens, proposal_probs, *_ = sampler_output_to_torch(sampler_output, sampler_transposed)
+
+        # Initialize tensors with appropriate sizes
+        entire_proposal_tokens = torch.full_like(proposal_tokens, -1, dtype=torch.long, device=self._device)
+        entire_proposal_probs = torch.zeros_like(proposal_probs, dtype=torch.float32, device=self._device)
```
Comment (njhill) [inline vllm/spec_decode/top1_proposer.py:242]: Is the first dim of these tensors always already `batch_size`? If so then this simplification makes sense!

File: vllm/spec_decode/top1_proposer.py [hunk 4]
```diff
@@ -226,47 +226,31 @@ def _merge_outputs(
...
-        entire_proposal_probs = proposal_probs.new_zeros(
-            batch_size,
-            *proposal_probs.shape[1:],
-        )
-        entire_proposal_probs[nonzero_proposal_len_indices] = proposal_probs
-
-        proposal_tokens, proposal_probs = (
-            entire_proposal_tokens,
-            entire_proposal_probs,
-        )
```
Comment (njhill) [inline vllm/spec_decode/top1_proposer.py:265]: Removing intermediate variables makes sense

**Review by njhill — COMMENTED** [review]
Thanks @jacob-crux! It looks like this mostly comes down to indexing using an on-device tensor rather than a list. This can also be used with the existing `tensor[index_tensor]` syntax ... are the `index_` tensor methods much faster?

**Comment by jacob-crux** [comment]
> Thanks @jacob-crux! It looks like this mostly comes down to indexing using an on-device tensor rather than a list. This can also be used with the existing `tensor[index_tensor]` syntax ... are the `index_` tensor methods much faster?

Hello!
First, I confirmed that when slicing with `output_indices_to_retain`, which is a `List[int]` in the `_filter_model_output` function, `cudaMemcpy` is used to copy and sync the index every time.

When comparing the speed difference between the list slice kernel and `torch.index_select` with the test code, I found that there was a speed difference of about 40-50% when comparing only the kernel. However, since the proportion of the existing kernel is small, it is difficult to get a big feeling.

Below are the profiling results. The first result is from the existing code, the second result is when only the index is assigned to a CUDA tensor, and the last result is from the code using `torch.index_select` in the PR.

File: vllm/spec_decode/top1_proposer.py (hunk 1 above)
Comment (jacob-crux) [reply vllm/spec_decode/top1_proposer.py:234]: I confirmed that the speed decrease occurs due to `cudamemcpy` in the process of generating tensor.
As you can see, the PR code has the same result, and you can get the effect of speed improvement.

```
if maybe_sampler_output is None:
    with nvtx.annotate("allocation"):
```
I set up nvtx as above and checked the profiling result below.
Above is the previous code, below is the PR code.

File: vllm/spec_decode/multi_step_worker.py (hunk 2 above)
Comment (jacob-crux) [reply vllm/spec_decode/multi_step_worker.py:215]: Please refer to the speed difference between list slicing and `tensor.index_select` that I have profiled. 
Since this part does not significantly impact performance, we can implement the change by simply modifying the `if statement` and converting `output_indices_to_retain` to a tensor, as you mentioned.

File: vllm/spec_decode/top1_proposer.py (hunk 3 above)
Comment (jacob-crux) [reply vllm/spec_decode/top1_proposer.py:242]: After checking several cases of speculative decoding, I found that when ngram was used, the first dim was not the batch size, so I restored the code for that part.

**Comment by njhill** [comment]
Thanks @jacob-crux, I wasn't questioning that using repeated list-based slicing is slower. I was suggesting that the same syntax can be  used to slice using a tensor. Then the actual code modifications needed would be minimal.

i.e.
```python
index_list = [1, 3, 5]
index_tensor = torch.tensor(index_list, device=device)
sliced = a_tensor[index_tensor]
sliced2 = another_tensor[index_tensor]
```

**Comment by jacob-crux** [comment]
> Thanks @jacob-crux, I wasn't questioning that using repeated list-based slicing is slower. I was suggesting that the same syntax can be used to slice using a tensor. Then the actual code modifications needed would be minimal.
> 
> i.e.
> 
> ```python
> index_list = [1, 3, 5]
> index_tensor = torch.tensor(index_list, device=device)
> sliced = a_tensor[index_tensor]
> sliced2 = another_tensor[index_tensor]
> ```

@njhill Thank you for your review.
[The comment](https://github.com/vllm-project/vllm/pull/7875#issuecomment-2311657316) was an experiment with the code you mentioned.

Here is a simple code example of the 3 profiled contents in the picture in the comment.

```python
# profile 1
index_list = [1, 3, 5]
sliced = a_tensor[index_tensor]

# profile 2
index_list = [1, 3, 5]
index_tensor = torch.tensor(index_list, device=device)
sliced = a_tensor[index_tensor]

# profile 3
index_list = [1, 3, 5]
index_tensor = torch.tensor(index_list, device=device)
sliced = a_tensor.index_select(0, index_tensor)
```
The difference mentioned here is
- The difference between profile 1 and profile 2 is that memcpy occurs frequently.
- The difference between profile 1&profile 2 and profile 3 is the difference in the GPU kernel called when slicing, and as in the middle section of the picture, the kernel shows a speed difference of about 40~50%.

Although the code of profile 2 may look a little simpler, the kernel of profile 3 has the effect of improving performance on the GPU.

**Comment by mgoin** [comment]
Closed as stale for now.
